nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,372 @@
+  node->get_parameter(name_ + "." + "speed_limit_topic", speed_limit_topic_);
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    node->get_logger(),
+    "Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic_.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::filterInfoCallback, this, std::placeholders::_1));
+
+  // Get global frame required for speed limit publisher
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  // Create new speed limit publisher
+  speed_limit_pub_ = node->create_publisher<nav2_msgs::msg::SpeedLimit>(
+    speed_limit_topic_,
+    rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable());
+  speed_limit_pub_->on_activate();
+
+  // Reset speed conversion states
+  base_ = BASE_DEFAULT;
+  multiplier_ = MULTIPLIER_DEFAULT;
+  percentage_ = true;
+  speed_limit_ = NO_SPEED_LIMIT;
+  speed_limit_prev_ = FAKE_SPEED_LIMIT;  // Fake speed limit to force speed limit topic update
+}
+
+void SpeedFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  // Set base_/multiplier_ or use speed limit in % of maximum speed
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  if (msg->type == SPEED_FILTER_ABSOLUTE) {
+    // Using speed limit in m/s
+    percentage_ = false;
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Speed filter is using absolute speed_limit = %f + filter_mask_data * %f",
+      base_, multiplier_);
+  } else if (msg->type == SPEED_FILTER_PERCENT) {
+    // Using speed limit in % of maximum speed
+    percentage_ = true;
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Speed filter is using expressed in a percent from maximum speed"
+      "speed_limit = %f + filter_mask_data * %f",
+      base_, multiplier_);
+  } else {
+    RCLCPP_ERROR(node->get_logger(), "Mode is not supported by SpeedFilter");
+    throw std::runtime_error("Mode is not supported by SpeedFilter");
+  }
+
+  // Reset speed limit when got new filter info
+  speed_limit_ = NO_SPEED_LIMIT;
+  speed_limit_prev_ = FAKE_SPEED_LIMIT;  // Fake speed limit to force speed limit topic update
```

Comment:
Why is `0` not good enough?

Reply:
Removed FAKE_SPEED_LIMIT here: https://github.com/ros-planning/navigation2/pull/2074/commits/3465c90415afd11418f98fc01bbc9386afa853ea#diff-6f9b5b73d858cc4570fd2d807b4ab6ad2c43be02da2ec6f34db8fac5708b9a50L56-R57 and assigning default values there: https://github.com/ros-planning/navigation2/pull/2074/commits/3465c90415afd11418f98fc01bbc9386afa853ea#diff-d344e376ecbc8e5ecd5045731e13855c6d2a9a955cdaee9b31dddbbee7681466R53